evap/staff/views.py

```diff
@@ -2226,12 +2227,24 @@ def user_bulk_update(request):
 def user_merge_selection(request):
     form = UserMergeSelectionForm(request.POST or None)
 
+    all_users = list(UserProfile.objects.all())
+    suggested_merges = []
+    for user1 in all_users:
+        if user1 != None and user1.is_external == False:
+            for user2 in all_users:
```

Comment:
As you already indicated, we'll have to do something a bit more efficient here.

@niklasmohrin Postgres has string functions that would allow us the email-to-username-portion conversion (e.g. `split_part`), we could then use database joining to annotate users with a merge candidate, and only return those users that actually have one. However, the database function is unique to postgres. We could also manually do a hash join in python, this would be portable across databases. Preferences?

Reply:
django code that finds a merge candidate in the database:
```python3
    from django.db.models import Func, Value, CharField, Subquery
    from evap.evaluation.models import UserProfile

    def email_to_local_part(field):
        return Func(field, Value('"@"'), Value(1), output_field=CharField(), function="split_part")

    query = UserProfile.objects.annotate(email_user_part=email_to_local_part("email"))
    query = query.annotate(
        merge_candidate_pk=Subquery(
            query.filter(pk__gt=OuterRef("pk"), email_user_part=email_to_local_part(OuterRef("email"))).values("pk")[:1]
        )
    )
    
    # -- untested from here --
    # not sure if we can make django directly annotate "into" a foreign key field so we don't have to manually do this
    users_with_merge_candidates = list(query.filter(merge_candidate_pk__isnull=False))
    merge_candidate_by_pk = {(user.pk, user) for user in UserProfile.objects.filter(
        pk__in=[inner_user.merge_candidate_pk for inner_user in users_with_merge_candidates]
    )}
    
    for user in users_with_merge_candidate:
        user.merge_candidate = merge_candidates_by_pk[user.merge_candidate_pk].
        
    # and then, it should be very easy, as you can just use
    print(user.merge_candidate)
    # to access a the merge candidate of each user.
```